File: week7/Maze_DFS.java
package week7;

import java.io.*;
import java.util.*;

/**
 * 미로탈출 - DFS
 * 동빈이는N×M크기의직사각형형태의미로에갇혔습니다.미로에는여러마리의괴물이있 어 이를 피해 탈출해야 합니다.
 * 동빈이의위치는(1,1)이며미로의출구는(N,M)의위치에존재하며한번에한칸씩이동할 수 있습니다. 이때 괴물이 있는 부분은 0으로, 괴물이 없는 부분은 1로 표시되어 있습니다. 미로 는 반드시 탈출할 수 있는 형태로 제시됩니다.
 * 이때동빈이가탈출하기위해움직여야하는최소칸의개수를구하세요.칸을셀때는시작칸 과 마지막 칸을 모두 포함해서 계산합니다.
 * 
 * 출처) https://www.acmicpc.net/problem/2178
 * 
 */

public class Maze_DFS {
    public static int N, M;
    public static int[][] graph = new int[200][200]; // 이동할 네 가지 방향 정의{상,하,좌,우}
    public static int dx[] = { -1, 1, 0, 0 };
    public static int dy[] = { 0, 0, -1, 1 };

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in); // N,M 입력
        N = scanner.nextInt();
        M = scanner.nextInt();
        scanner.nextLine(); // 버퍼 지우기 System.out.println("n " + N + "m" + M);
        // 2차원 배열의 정보 입력 받기
        for (int i = 0; i < N; i++) {
            String str = scanner.nextLine();
            for (int j = 0; j < M; j++) {
                graph[i][j] = str.charAt(j) - '0';
                scanner.close();
                System.out.println("result = " + dfs(0, 0));
            }
        }
    }

    private static int dfs(int x, int y) {
        Stack<Node> stack = new Stack<Node>();
        Node node = new Node(x, y);
        stack.push(node);
        while (!stack.empty()) {
            node = stack.pop();
            x = node.getX();
            y = node.getY();
            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];
                // 미로 범위를 벗어나면 무시
                if (nx < 0 || ny < 0 || nx >= N || ny >= M)
                    continue;
                // 한 번 왔던 위치면 무시,괴물이 있는 위치면 무시 
                if (graph[nx][ny]==1 ){
                    System.out.println("[" + nx + "," + ny + "]");
                    graph[nx][ny] = graph[x][y] + 1;
                    stack.push(new Node(nx, ny));
                }
            }
        }   
        return graph[N-1][M-1];
    }
 
    static class Node {
        final private int x;
        final private int y;

        Node(int x, int y) {
            this.x = x;
            this.y = y;
        }

        public int getX() {
            return x;
        }

        public int getY() {
            return y;
        }

    }

}

File: week7/maze_DFS.py
'''
미로탈출 - DFS
동빈이는N×M크기의직사각형형태의미로에갇혔습니다.미로에는여러마리의괴물이있 어 이를 피해 탈출해야 합니다.
동빈이의위치는(1,1)이며미로의출구는(N,M)의위치에존재하며한번에한칸씩이동할 수 있습니다. 이때 괴물이 있는 부분은 0으로, 괴물이 없는 부분은 1로 표시되어 있습니다. 미로 는 반드시 탈출할 수 있는 형태로 제시됩니다.
이때동빈이가탈출하기위해움직여야하는최소칸의개수를구하세요.칸을셀때는시작칸 과 마지막 칸을 모두 포함해서 계산합니다.

출처) https://www.acmicpc.net/problem/2178
'''

from collections import deque

n, m = map(int, input().split())  # 행, 열
graph = []

# 데이터 받아오기
for i in range(n):
    graph.append(list(map(int, input())))

# 움직임에 따른 네 방향 정의(좌, 우, 상, 하)
dx = [-1, 1, 0, 0]  # 좌우로 움직임
dy = [0, 0, -1, 1]  # 위아래로 움직임


def bfs(x, y):
    queue = deque()  # 큐 구현을 위해 deque 라이브러리 사용
    queue.append((x, y))

    while queue:
        x, y = queue.popleft()  # 기존값 초기화
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
        #  미로 찾기 공간을 벗어난 경우 무시
        if nx < 0 or ny < 0 or nx >= n or ny >= m:
            continue
        # 움직일 수 없는 공간(벽)일 경우 무시
        if graph[nx][ny] == 0:
            continue
        # 노드를 처음 방문하는 경우
        if graph[nx][ny] == 1:
            graph[nx][
                ny] = graph[x][y] + 1  # 기존 좌표(노드)의 값에 +1 한 값을 움직인 좌표(노드)에 추가
            queue.append((nx, ny))

    return graph[n - 1][m - 1]


print(bfs(0, 0))